### PR TITLE
Speed up decoding of large enums

### PR DESCRIPTION
**Pull request by zyla**
Generic `to` is quite slow on datatypes with many constructors, because it runs `O(log^2(n))` `instanceof` operations, where `n` is the number of constructors.

This change moves this cost from the decoding loop to initialization time. Now we store values of the target type in the lookup table instead of Generic representations. This makes initialization slower, but if there are many enums values to decode, we save time there. 

Before:

```
Enum3 Unscramble                        : 0.000027ms/op
Enum10 Unscramble                       : 0.000169ms/op
Enum30 Unscramble                       : 0.002258ms/op
```

After:

```
Enum3 Unscramble                        : 0.000016ms/op
Enum10 Unscramble                       : 0.000047ms/op
Enum30 Unscramble                       : 0.000050ms/op
```

As seen in the above benchmark results, after this change enum decoding is almost independent of enum size. Which makes sense, because now it's only a hashtable lookup, as it should be.

### CODE REVIEW

File: src/Unscramble/Enum.purs [hunk 1]
```diff
@@ -2,42 +2,42 @@ module Unscramble.Enum where
...
 import Foreign.Object as Object
 
-genericUnsafeDecodeEnum :: forall a rep. Generic a rep => EnumConstructors rep => EnumOptions -> Foreign -> a
+genericUnsafeDecodeEnum :: forall a rep. Generic a rep => EnumConstructors rep a => EnumOptions -> Foreign -> a
 genericUnsafeDecodeEnum opts =
-  let constructors = Object.fromFoldable (enumConstructors opts :: Array (Tuple String rep))
+  let constructors = Object.fromFoldable (enumConstructors to opts :: Array (Tuple String a))
   in \value ->
     let tag = decodeString value in
     case Object.lookup tag constructors of
```
Comment (kozak) [inline src/Unscramble/Enum.purs:19]: When you say about the hash lookup, this is what you mean (not a lookup in instance dict) ?

**Comment by zyla** [comment]
To clarify: the `instanceof`s come from the generated Generic instances. For example, here is `to` from the instance for Enum10:

```
    to: function (x) {
        if (x instanceof Data_Generic_Rep.Inl) {
            return E10_1.value;
        };
        if (x instanceof Data_Generic_Rep.Inr && x.value0 instanceof Data_Generic_Rep.Inl) {
            return E10_2.value;
        };
        if (x instanceof Data_Generic_Rep.Inr && (x.value0 instanceof Data_Generic_Rep.Inr && x.value0.value0 instanceof Data_Generic_Rep.Inl)) {
            return E10_3.value;
        };
        if (x instanceof Data_Generic_Rep.Inr && (x.value0 instanceof Data_Generic_Rep.Inr && (x.value0.value0 instanceof Data_Generic_Rep.Inr && x.value0.value0.value0 instanceof Data_Generic_Rep.Inl))) {
            return E10_4.value;
        };
        if (x instanceof Data_Generic_Rep.Inr && (x.value0 instanceof Data_Generic_Rep.Inr && (x.value0.value0 instanceof Data_Generic_Rep.Inr && (x.value0.value0.value0 instanceof Data_Generic_Rep.Inr && x.value0.value0.value0.value0 instanceof Data_Generic_Rep.Inl)))) {
            return E10_5.value;
        };
        if (x instanceof Data_Generic_Rep.Inr && (x.value0 instanceof Data_Generic_Rep.Inr && (x.value0.value0 instanceof Data_Generic_Rep.Inr && (x.value0.value0.value0 instanceof Data_Generic_Rep.Inr && (x.value0.value0.value0.value0 instanceof Data_Generic_Rep.Inr && x.value0.value0.value0.value0.value0 instanceof Data_Generic_Rep.Inl))))) {
            return E10_6.value;
        };
        if (x instanceof Data_Generic_Rep.Inr && (x.value0 instanceof Data_Generic_Rep.Inr && (x.value0.value0 instanceof Data_Generic_Rep.Inr && (x.value0.value0.value0 instanceof Data_Generic_Rep.Inr && (x.value0.value0.value0.value0 instanceof Data_Generic_Rep.Inr && (x.value0.value0.value0.value0.value0 instanceof Data_Generic_Rep.Inr && x.value0.value0.value0.value0.value0.value0 instanceof Data_Generic_Rep.Inl)))))) {
            return E10_7.value;
        };
        if (x instanceof Data_Generic_Rep.Inr && (x.value0 instanceof Data_Generic_Rep.Inr && (x.value0.value0 instanceof Data_Generic_Rep.Inr && (x.value0.value0.value0 instanceof Data_Generic_Rep.Inr && (x.value0.value0.value0.value0 instanceof Data_Generic_Rep.Inr && (x.value0.value0.value0.value0.value0 instanceof Data_Generic_Rep.Inr && (x.value0.value0.value0.value0.value0.value0 instanceof Data_Generic_Rep.Inr && x.value0.value0.value0.value0.value0.value0.value0 instanceof Data_Generic_Rep.Inl))))))) {
            return E10_8.value;
        };
        if (x instanceof Data_Generic_Rep.Inr && (x.value0 instanceof Data_Generic_Rep.Inr && (x.value0.value0 instanceof Data_Generic_Rep.Inr && (x.value0.value0.value0 instanceof Data_Generic_Rep.Inr && (x.value0.value0.value0.value0 instanceof Data_Generic_Rep.Inr && (x.value0.value0.value0.value0.value0 instanceof Data_Generic_Rep.Inr && (x.value0.value0.value0.value0.value0.value0 instanceof Data_Generic_Rep.Inr && (x.value0.value0.value0.value0.value0.value0.value0 instanceof Data_Generic_Rep.Inr && x.value0.value0.value0.value0.value0.value0.value0.value0 instanceof Data_Generic_Rep.Inl)))))))) {
            return E10_9.value;
        };
        if (x instanceof Data_Generic_Rep.Inr && (x.value0 instanceof Data_Generic_Rep.Inr && (x.value0.value0 instanceof Data_Generic_Rep.Inr && (x.value0.value0.value0 instanceof Data_Generic_Rep.Inr && (x.value0.value0.value0.value0 instanceof Data_Generic_Rep.Inr && (x.value0.value0.value0.value0.value0 instanceof Data_Generic_Rep.Inr && (x.value0.value0.value0.value0.value0.value0 instanceof Data_Generic_Rep.Inr && (x.value0.value0.value0.value0.value0.value0.value0 instanceof Data_Generic_Rep.Inr && x.value0.value0.value0.value0.value0.value0.value0.value0 instanceof Data_Generic_Rep.Inr)))))))) {
            return E10_10.value;
        };
        throw new Error("Failed pattern match at Bench.Micro (line 144, column 1 - line 144, column 33): " + [ x.constructor.name ]);
    },
```

Also, the complexity I provided in PR description is wrong. The generic tree seems to be unbalanced, to it's `O(n^2)`, not `O(log^2(n))`.

File: src/Unscramble/Enum.purs (hunk 1 above)
Comment (zyla) [reply src/Unscramble/Enum.purs:19]: Yes (assuming JS `Object` is a hashtable, which it will probably be in this case in most implementations)

**Review by kozak — APPROVED** [review]
If my understanding (in the comment) is ok, then lg to me :)

**Pull request merged**